pages/en/lb4/Testing-Your-application.md

```diff
@@ -8,39 +8,236 @@ permalink: /doc/en/lb4/Testing-your-application.html
+
+For example, if our `Person` Model has properties `firstname`, `middlename` and `surname` and provides a function to obtain the full name, then we should write unit tests to verify the implementation of this additional method.
+
+```ts
+// test/unit/models/person.model.unit.ts
+import {Person} from '../../models/person.model.ts'
+import {expect} from '@loopback/testlab';
+
+describe('Person (unit)', () => {
+  // we recommend to group tests by method names
+  describe('getFullName()', () => {
+    it('uses all three parts when present', () => {
+      const person = new Person({
+        firstname: 'Jane',
+        middlename: 'Smith',
+        surname: 'Brown'
+      });
+
+      const fullName = person.getFullName();
+
+      expect(fullName).to.equal('Jane Smith Brown');
+    });
+
+    it('omits middlename when not present', () => {
+      // etc.
+    });
+  });
+});
+```
+
+Writing a unit test for a custom Repository methods is not straightforward because `CrudRepositoryImpl` is based on legacy loopback-datasource-juggler that was not designed with Dependency Injection in mind. We are recommending to use integration tests to verify the implementation of custom Repository methods, see [Test your Repositories against a real database](#test-your-repositories-against-a-real-database) in [Integration Testing](#integration-testing).
+
+### Unit-test your Sequence
+
+While it's possible to test a custom Sequence class in isolation, we think it's better to rely on acceptance-level tests in this exceptional case. The reason is that a custom Sequence class typically has many dependencies (which makes test setup too long and complex), and at the same time it provides very little functionality on top of the injected sequence actions. Bugs are much more likely to caused by the way how the real sequence action implementations interact together (which is not covered by unit tests), instead of the sequence code itself (which is the only thing covered).
+
+See [Test Sequence customizations](#test-sequence-customizations) in [Acceptance Testing](#acceptance-testing).
+
+### Unit-test your Services
+
+To be done. The initial Beta release does not include Services as a first-class feature. The following GitHub issues are tracking the related work:
+
+ - Define services to represent interactions with REST APIs, SOAP Web Services, gRPC services, and more: [#522](https://github.com/strongloop/loopback-next/issues/522)
+ - Guide: Services [#451](https://github.com/strongloop/loopback-next/issues/451)
+
+## Integration testing
+
+Integration tests are considered as "white-box" too. They use "inside-out" approach that tests how multiple units work together or with external services. Test doubles may be used to isolate tested units from external variables/state that's not part of the tested scenario.
```

Comment:
`Integration tests are considered as "white-box" too.` --> `Integration tests are also considered "white-box" tests.`

Reply:
`They use "inside-out" approach that tests how multiple units work together or with external services.` -> `They use the "inside-out" approach, which tests how multiple units work together or with external services.`